### PR TITLE
Base: Added Grid System Processor

### PR DESCRIPTION
**Pull request by r-englund**
DEMO: https://www.youtube.com/watch?v=s79T3Seo5jM

grid

TODO before ready for review
- [x] Check/Add example workspaces
- [x] Add regression tests
- [x] Check documentation (Code / Processors doc)

### CODE REVIEW

File: modules/base/include/modules/base/processors/gridsystem.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+template <unsigned N, typename T, typename Indices = std::make_index_sequence<N>>
+constexpr std::array<T, N> copyInitFilledArray(const T& t) {
+    return copyInitFilledArrayHelper(t, Indices{});
+}
+}  // namespace detail
+
+enum class ListOrSingleValuePropertyState { Single, List };
+
+template <typename Prop, unsigned N>
+class ListOrSingleValueProperty : public CompositeProperty {
```
Comment (r-englund) [inline modules/base/include/modules/base/processors/gridsystem.h]: Im not sure what to name this property

File: modules/base/include/modules/base/processors/gridsystem.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    virtual ~ListOrSingleValueProperty() = default;
+
+    TemplateOptionProperty<State> state_;
+    Prop single_;
+    std::array<Prop, N> list_;
+};
+
+/** \docpage{org.inviwo.GridSystem, Grid System}
+ * ![](org.inviwo.GridSystem.png?classIdentifier=org.inviwo.GridSystem)
+ * Explanation of how to use the processor.
```
Comment (martinfalk) [inline modules/base/include/modules/base/processors/gridsystem.h]: fill in the details

**Review by martinfalk — DISMISSED** [review]

File: modules/base/include/modules/base/processors/gridsystem.h [hunk 3]
```diff
@@ -41,20 +41,18 @@
...
-template <unsigned N, typename T>
-constexpr T passthrough_the_value_helper(const T& t) {
-    return t;
-}
 
-template <typename T, unsigned... I>
-constexpr auto copyInitFilledArrayHelper(const T& t, std::index_sequence<I...>) {
-    return std::array<T, sizeof...(I)>{passthrough_the_value_helper<I>(t)...};
+template <typename T, std::size_t... Is>
+constexpr std::array<T, sizeof...(Is)> create_array(T value, std::index_sequence<Is...>) {
```
Comment (petersteneteg) [inline modules/base/include/modules/base/processors/gridsystem.h]: we have util::make_array in stdextensions.h

**Review by petersteneteg — DISMISSED** [review]

File: modules/base/include/modules/base/processors/gridsystem.h (hunk 3 above)
Comment (r-englund) [reply modules/base/include/modules/base/processors/gridsystem.h]: Nice, using the `util::make_array` made the code a bit cleaner. (Not using copy ctor, not changing the dispName/id afterwards)

File: modules/base/include/modules/base/processors/gridplanes.h [hunk 4]
```diff
@@ -101,35 +101,40 @@ class ListOrSingleValueProperty : public CompositeProperty {
     std::array<Prop, N> list_;
 
 private:
-    const static std::array<std::string, 4> xyzwAxisNames = {"x", "y", "z", "w"};
+    const static inline std::array<std::string, 4> xyzwAxisNames = {"x", "y", "z", "w"};
```
Comment (petersteneteg) [inline modules/base/include/modules/base/processors/gridplanes.h]: can use static constexpr std::array<std::string_view, 4> ...

**Review by martinfalk — DISMISSED** [review]

**Review by martinfalk — APPROVED** [review]

**Pull request merged**